test/test_model.py

```diff
@@ -755,6 +759,223 @@ def assertBackendCalls(self, expected, *, reset=True):  # noqa: N802
         self.assertEqual(expected, self.harness._get_backend_calls(reset=reset))
 
 
+class PushPullCase:
+    """Test case for table-driven tests."""
+
+    def __init__(self, **vals):
+        self.pattern = None
+        self.dst = None
+        self.errorlist = []
+        self.wantlist = set()
+        for key in vals:
+            setattr(self, key, vals[key])
+
+
+recursive_list_cases = [
+    PushPullCase(
+        name='basic recursive list',
+        path='/',
+        files=['/foo/bar.txt', '/baz.txt'],
+        wantlist={'/foo/bar.txt', '/baz.txt'},
+    ),
+    PushPullCase(
+        name='basic recursive list reverse',
+        path='/',
+        files=['/baz.txt', '/foo/bar.txt'],
+        wantlist={'/baz.txt', '/foo/bar.txt'},
+    ),
+    PushPullCase(
+        name='directly list a (non-directory) file',
+        path='/baz.txt',
+        files=['/baz.txt'],
+        wantlist={'/baz.txt'},
+    ),
+]
+
+
+@pytest.mark.parametrize('case', recursive_list_cases)
```

Comment:
I like this approach very much - there are probably more places in the framework where this could be used to simplify our test suite.

Reply:
Yeah.  My time in the Go space really taught me to appreciate the table-driven testing style.